SpicesUpdate@claudiux/files/SpicesUpdate@claudiux/applet.js

```diff
@@ -0,0 +1,1097 @@
+      null);
+
+    this.settings.bindProperty(Settings.BindingDirection.IN,
+      "displayType",
+      "displayType",
+      this.on_settings_changed,
+      null);
+
+    this.on_orientation_changed(orientation);
+
+    // Init lists of Spices:
+    this.populateSettingsUnprotectedApplets();
+    this.populateSettingsUnprotectedDesklets();
+    this.populateSettingsUnprotectedExtensions();
+    this.populateSettingsUnprotectedThemes();
+
+    // Dependencies:
+    this.dependenciesMet = this.are_dependencies_installed();
+    if (!this.dependenciesMet) this.refreshInterval = 5;
+
+    // ++ Set up Left Click Menu
+    this.menuManager = new PopupMenu.PopupMenuManager(this);
+    this.menu = new Applet.AppletPopupMenu(this, orientation);
+    this.menuManager.addMenu(this.menu);
+
+    // Default icon color
+    //let themeNode = this.actor.get_theme_node();
+    //let icon_color = themeNode.get_icon_colors();
+    //this.defaultColor = icon_color.foreground.to_string();
+    this.defaultColor = "#000000FF";
+
+    // Monitoring metadata.json files
+    this.monitors = new Array();
+
+    // Count of Spices to update
+    this.nb_to_update = 0;
+
+    // Badge
+    this.badge = new St.BoxLayout({
+      style_class: 'grouped-window-list-badge',
+      important: true,
+      width: 12 * global.ui_scale,
+      height: 12 * global.ui_scale,
+      x_align: St.Align.MIDDLE,
+      y_align: St.Align.MIDDLE,
+      show_on_set_parent: false,
+      style: 'margin: 0;',
+    });
+    this.numberLabel = new St.Label({
+      style: 'font-size: 10px;padding: 0px;',
+      style_class: 'grouped-window-list-number-label',
+      important: true,
+      text: '',
+      anchor_x: -3 * global.ui_scale,
+      anchor_y: 1 + (global.ui_scale > 1 ? 2 : 0)
+    });
+    this.numberLabel.clutter_text.ellipsize = false;
+    this.badge.add(this.numberLabel, {
+      x_align: St.Align.START,
+      y_align: St.Align.START,
+    });
+    this.actor.add_child(this.badge);
+
+    this.testblink = [];
+    this.forceRefresh = false;
+    this.applet_running = true;
+    this.first_loop = true; // To do nothing for 1 minute.
+    this.on_settings_changed();
+  }; // End of constructor
+
+  on_orientation_changed (orientation) {
+    this.orientation = orientation;
+    this.isHorizontal = !(this.orientation == St.Side.LEFT || this.orientation == St.Side.RIGHT);
+    this._set_main_label();
+  }; // End of on_orientation_changed
+
+  _set_main_label() {
+    if (this.displayType === "compact") {
+      this.set_applet_label("");
+    } else {
+      if (this.isHorizontal === true) {
+        this.set_applet_label("Spices Update");
+      } else {
+        this.set_applet_label("SpU");
+      }
+    }
+  }; // End of _set_main_label
+
+  // ++ Function called when settings are changed
+  on_settings_changed() {
+    // Label
+    this._set_main_label();
+
+    // Refresh intervall:
+    this.refreshInterval = 3600*this.general_frequency;
```

Comment:
I'm hitting an issue with the settings. When you change a setting, it starts notifying and doesn't stop. In this case, I changed "Notify me by changing the icon when Spices need an update".

```
(cinnamon:8955): Cjs-CRITICAL **: 15:03:34.925: Object St.Button (0x560ad68e57a0), has been already finalized. Impossible to get any property from it.
== Stack trace for context 0x560a962704c0 ==
#0 0x560a96bc3b00 i   /usr/share/cinnamon/js/ui/messageTray.js:1796 (0x7f8975192918 @ 289)
#1 0x7ffe59ee0040 b   self-hosted:914 (0x7f89761e6670 @ 346)
Cjs-Message: 15:03:35.244: JS LOG: [LookingGlass/warning] Invalid or null source id used when attempting to run Mainloop.source_remove()
overrideDumpStack/global.dump_gjs_stack@/usr/share/cinnamon/js/ui/overrides.js:103:42
overrideMainloop/Mainloop.source_remove@/usr/share/cinnamon/js/ui/overrides.js:137:13
MessageTray.prototype._updateNotificationTimeout@/usr/share/cinnamon/js/ui/messageTray.js:1784:13
MessageTray.prototype._onNotificationDestroy@/usr/share/cinnamon/js/ui/messageTray.js:1528:13
_emit@resource:///org/cinnamon/cjs/modules/signals.js:142:27
Notification.prototype._onDestroy@/usr/share/cinnamon/js/ui/messageTray.js:1027:9
destroy@/usr/share/cinnamon/js/ui/environment.js:118:13
Notification.prototype.destroy@/usr/share/cinnamon/js/ui/messageTray.js:1032:9
Source.prototype.destroyNonResidentNotifications@/usr/share/cinnamon/js/ui/messageTray.js:1210:17
Source.prototype.open@/usr/share/cinnamon/js/ui/notificationDaemon.js:700:9
_emit@resource:///org/cinnamon/cjs/modules/signals.js:142:27
Notification.prototype._onClicked@/usr/share/cinnamon/js/ui/messageTray.js:1012:9
```

Maybe try stopping the loop when the settings change. When the timeout callback returns a falsey value, it stops the loop. Instead of it recursively starting a new loop at the end of `updateLoop`, you could try having it return a property on the instance, and change that instance to true/false depending on the situation. An example of that is in the [multicore-sys-monitor applet](https://github.com/linuxmint/cinnamon-spices-applets/blob/f5ba871afc589a75abf0825373ef3516356c3cac/multicore-sys-monitor%40ccadeptic23/files/multicore-sys-monitor%40ccadeptic23/4.0/applet.js#L331).

Reply:
I have modified the management of the loop. Is this better?

Reply:
Yep, LGTM. Thanks.